### PR TITLE
feat: Provide scaler for Amazon managed service for Prometheus

### PR DESCRIPTION
**Pull request by sguruvar**
_Provide a description of what has been changed_

### Checklist

- [x] When introducing a new scaler, I agree with the [scaling governance policy](https://github.com/kedacore/governance/blob/main/SCALERS.md)
- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Tests have been added
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Fixes https://github.com/kedacore/keda/issues/2214

### CODE REVIEW

**Comment by sguruvar** [comment]
Incorporated the changes with SDK v2.. Added the unit and e2e tests.. Spinned up a new AWS AMP instance and the flow works good too.. can someone from Keda reviewer team help merge this for closure?

File: pkg/scalers/aws_sigv4.go
Comment (JorTurFer) [inline pkg/scalers/aws_sigv4.go:103]: Do we need this? I mean, `metadata` already has `awsRegion` but also if you don't set it, the underlying client uses default region

File: pkg/scalers/aws_sigv4.go
Comment (JorTurFer) [inline pkg/scalers/aws_sigv4.go:67]: Is the region mandatory? AFAIK aws client has a default region that uses when region is not provided

File: pkg/scalers/aws_sigv4_test.go
Comment (JorTurFer) [inline pkg/scalers/aws_sigv4_test.go:16]: Remove this comment please

File: tests/scalers/aws/aws_managed_prometheus_pod_identity/aws_managed_prometheus_pod_identity_test.go
Comment (JorTurFer) [inline tests/scalers/aws/aws_managed_prometheus_pod_identity/aws_managed_prometheus_pod_identity_test.go:49]: Auths `aws-eks` and `aws-kiam` are deprecated. Although testing them it's nice, we have to test mandatorily `aws` auth (and we can add `aws-eks` too if we want)
```suggestion
    provider: aws
```

File: pkg/scalers/aws_sigv4.go
Comment (JorTurFer) [inline pkg/scalers/aws_sigv4.go:21]: Is this in use?

File: pkg/scalers/aws_sigv4_test.go
Comment (JorTurFer) [inline pkg/scalers/aws_sigv4_test.go:9]:
```suggestion
	"github.com/stretchr/testify/require"

	"github.com/kedacore/keda/v2/pkg/util"
```

**Comment by JorTurFer** [comment]
Also, the changelog has to been updated

**Comment by JorTurFer** [comment]
And last but not least, please fix DCO check. You can get more info about how to solve it clicking in the check link
image

File: pkg/scalers/aws_sigv4.go
Comment (zroubalik) [inline pkg/scalers/aws_sigv4.go:1]: we have a package for aws related stuff, we shoudl probably move it there: `scalers/aws`

**Review by zroubalik — COMMENTED** [review]
We should also update documentation for 2.13: https://github.com/kedacore/keda-docs

**Comment by sguruvar** [comment]
> You have resolved the conversations but after the push I still see the code changes. Maybe you didn't commit the change by error

working on it.. should be closed in sometime

File: CHANGELOG.md
Comment (JorTurFer) [inline CHANGELOG.md:123]:
```suggestion
- **Prometheus Metrics**: Introduce paused ScaledObjects in Prometheus metrics ([#4430](https://github.com/kedacore/keda/issues/4430))
- **Prometheus Scaler**: Provide scaler for Amazon managed service for Prometheus ([#2214](https://github.com/kedacore/keda/issues/2214))
- **Pulsar Scaler**: support endpointParams in pulsar oauth ([#5069](https://github.com/kedacore/keda/issues/5069))
```

File: pkg/scalers/aws_sigv4.go
Comment (JorTurFer) [inline pkg/scalers/aws_sigv4.go:93]:
```suggestion
	awsCfg, err := awsutils.GetAwsConfig(context.Background(), metadata.awsRegion, metadata.awsAuthorization)
	if err != nil {
		return nil, err
	}

	client := amp.NewFromConfig(*awsCfg, func(o *amp.Options) {})
	rt := &roundTripper{
		client: client,
		region: metadata.awsRegion,
	}
```

**Review by JorTurFer — COMMENTED** [review]
Looking good. From my side, the PR is ready to merge once the moving of `aws_sigv4.go` and `aws_sigv4_test.go` files is done

File: pkg/scalers/aws_sigv4.go
Comment (JorTurFer) [reply pkg/scalers/aws_sigv4.go:1]: This is still pending

**Comment by sguruvar** [comment]
Doc PR: https://github.com/kedacore/keda-docs/pull/1291

**Comment by zroubalik** [comment]
> > Looking good. From my side, the PR is ready to merge once the moving of `aws_sigv4.go` and `aws_sigv4_test.go` files is done
> 
> changes done..moving aws_sigv4 and its test to aws directory creates circular dependency.. i.e. Prometheus_scaler calls NewSigV4RoundTripper from aws. Whilst from aws_sigv4 uses the ScalerConfig from scaler again. i.e :
> 
> pkg/scalers/prometheus_scaler ===> pkg/scalers/aws/aws_sigv4 ===> pkg/scalers/scaler <=== pkg/scalers/prometheus_scaler

Yeah, ScalersConfig is the problem, we should refactor this. So let's proceed with the current state for now.

@sguruvar there are some unused imports, see the reported validation problems.

File: pkg/scalers/aws_sigv4.go
Comment (zroubalik) [reply pkg/scalers/aws_sigv4.go:1]: There is a problem with ScalersConfig circular dependency, we should refactor the package: https://github.com/kedacore/keda/issues/5379

**Comment by zroubalik** [comment]
@sguruvar I refactored the codebase, ScalersConfig is in different package, please refactor this PR and move the new AWS code to `pkg/scalers/aws` package.
See: https://github.com/kedacore/keda/pull/5381

**Comment by sguruvar** [comment]
> We should also update documentation for 2.13: https://github.com/kedacore/keda-docs

Done

**Comment by sguruvar** [comment]
@zroubalik Files moved to aws folder, Changes done and pushed

File: pkg/scalers/aws/aws_sigv4.go
Comment (zroubalik) [reply pkg/scalers/aws/aws_sigv4.go:36]: What about this problem? Nothing has been written to hasher, so it's the hash of an empty string, which is strange 🤷‍♂️

**Comment by zroubalik** [comment]
Please also rebase, we included some aws related fixes: https://github.com/kedacore/keda/pull/5389

File: pkg/scalers/aws/aws_sigv4.go
Comment (JorTurFer) [reply pkg/scalers/aws/aws_sigv4.go:36]: Then let's modify it to use and empty string (when I helped to @sguruvar I didn't check it well and maybe it works, I'm checking a playground right now)

File: pkg/scalers/aws/aws_sigv4.go
Comment (JorTurFer) [inline pkg/scalers/aws/aws_sigv4.go:72]: As I've written by slack, this logic requires that AWS auth info is always available. This is not a problem for AWS Scalers, but as we are using this round tripper for adding AWS support on not AWS scalers, I think that we should assume that is can fail and hide the error here with a comment. Something like:
```suggestion
	// parseAwsAMPMetadata can return an error if AWS info is missing
	// but this can happen if we check for them on not AWS scalers
	// which is probably the reason to create a SigV4RoundTripper.
	// To prevent failures we check if the metadata is nil
	// (missing AWS info) and we hide the error
	metadata, _ := parseAwsAMPMetadata(config)
	if metadata == nil {
		return nil, nil
	}
```

File: pkg/scalers/aws/aws_sigv4.go
Comment (JorTurFer) [reply pkg/scalers/aws/aws_sigv4.go:36]: This should be equivalent and we'd not need to ignore the rule
```suggestion
	hasher := sha256.New()
	reqCxt := v4.SetPayloadHash(req.Context(), hex.EncodeToString(hasher.Sum([]byte{})))
```
WDYT @zroubalik ?

File: pkg/scalers/aws/aws_sigv4.go
Comment (sguruvar) [reply pkg/scalers/aws/aws_sigv4.go:36]: The SDK expect it to set an integrity check value on the request payload

Comment (zroubalik) [reply pkg/scalers/aws/aws_sigv4.go:36]: Can you link the relevant SDK docs here please?

File: pkg/scalers/aws/aws_sigv4.go
Comment (JorTurFer) [inline pkg/scalers/aws/aws_sigv4.go:37]: I think that this also works
```suggestion
	reqHash := hex.EncodeToString(hasher.Sum([]byte{}))
```

File: pkg/scalers/aws/aws_sigv4.go
Comment (JorTurFer) [inline pkg/scalers/aws/aws_sigv4.go:38]: I think that this will work too, and it avoid the requirement of using the hasher for hashing nothing
```suggestion
	// Sign request
	err = rt.client.Options().HTTPSignerV4.SignHTTP(req.Context(), cred, req, "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855", "aps", rt.region, time.Now())
```
Probably we could generate a const with the value and use it here for be clearer, but it can work.
I'd say that adding a comment claifying that "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855" is the sha256 of "" can be worth too

Comment (zroubalik) [reply pkg/scalers/aws/aws_sigv4.go:38]: no magic constants please :)

Comment (JorTurFer) [reply pkg/scalers/aws/aws_sigv4.go:38]: do you prefer to set the value inline?

File: pkg/scalers/aws/aws_sigv4.go
Comment (sguruvar) [reply pkg/scalers/aws/aws_sigv4.go:36]: https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/amp

File: pkg/scalers/aws/aws_sigv4.go
Comment (sguruvar) [reply pkg/scalers/aws/aws_sigv4.go]: when we introduce as below, the semgrp issue is

> // This is to avoid the semgrp issue of odd hash.Sum call flow when using hasher.sum(nil)
> 	// "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855" is the sha256 of ""
> 	const reqHash = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
> 
> 	err = rt.client.Options().HTTPSignerV4.SignHTTP(req.Context(), cred, req, reqHash, "aps", rt.region, time.Now())


Otherwise we could have it as above and ignore the semgrp issue

> hasher := sha256.New()
> reqCxt := v4.SetPayloadHash(req.Context(), hex.EncodeToString(hasher.Sum([]byte{})))
> reqHash := v4.GetPayloadHash(reqCxt)
> err =rt.client.Options().HTTPSignerV4.SignHTTP(req.Context(), cred, req, reqHash, "aps", rt.region, time.Now())

Please let me know the option to take here

File: tests/scalers/aws/aws_managed_prometheus_pod_identity/aws_managed_prometheus_pod_identity_test.go
Comment (JorTurFer) [inline tests/scalers/aws/aws_managed_prometheus_pod_identity/aws_managed_prometheus_pod_identity_test.go:126]:
```suggestion
	workspaceOutput, err := ampClient.CreateWorkspace(context.Background(), nil)
	assert.Error(t, err, "aws prometheus workspace creation has failed")
```

File: tests/scalers/aws/aws_managed_prometheus/aws_managed_prometheus_test.go
Comment (JorTurFer) [inline tests/scalers/aws/aws_managed_prometheus/aws_managed_prometheus_test.go:141]:
```suggestion
	workspaceOutput, err := ampClient.CreateWorkspace(context.Background(), nil)
	assert.Error(t, err, "aws prometheus workspace creation has failed")
```

**Review by JorTurFer — COMMENTED** [review]
new e2e are failing (with a panic) and I'd say that it's happening at workspace creation time

File: tests/scalers/aws/aws_managed_prometheus/aws_managed_prometheus_test.go
Comment (JorTurFer) [inline tests/scalers/aws/aws_managed_prometheus/aws_managed_prometheus_test.go:159]:
```suggestion
	_, err = ampClient.DeleteWorkspace(context.Background(), input)
```

File: tests/scalers/aws/aws_managed_prometheus_pod_identity/aws_managed_prometheus_pod_identity_test.go
Comment (JorTurFer) [inline tests/scalers/aws/aws_managed_prometheus_pod_identity/aws_managed_prometheus_pod_identity_test.go:144]:
```suggestion
	_, err = ampClient.DeleteWorkspace(context.Background(), input)
```

**Comment by JorTurFer** [comment]
Error logs from e2e test are really weird:
```bash
=== RUN   TestScaler
    aws_managed_prometheus_test.go:138: --- setting up ---
--- FAIL: TestScaler (0.42s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x10f5ac8]

goroutine 132 [running]:
testing.tRunner.func1.2({0x126caa0, 0x2327760})
	/usr/local/go/src/testing/testing.go:1545 +0x1c4
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1548 +0x360
panic({0x126caa0?, 0x2327760?})
	/usr/local/go/src/runtime/panic.go:914 +0x218
command-line-arguments_test.TestScaler(0x4000475860)
	/__w/keda/keda/tests/scalers/aws/aws_managed_prometheus/aws_managed_prometheus_test.go:143 +0x178
testing.tRunner(0x4000475860, 0x15808b0)
	/usr/local/go/src/testing/testing.go:1595 +0xe8
created by testing.(*T).Run in goroutine 1
	/usr/local/go/src/testing/testing.go:1648 +0x33c
FAIL	command-line-arguments	0.491s
FAIL
```

It looks that the test can't start for any reason, but the issue looks in the test script side, not in KEDA. Could you take a look?

**Comment by sguruvar** [comment]
> Error logs from e2e test are really weird:
> 
> ```shell
> === RUN   TestScaler
>     aws_managed_prometheus_test.go:138: --- setting up ---
> --- FAIL: TestScaler (0.42s)
> panic: runtime error: invalid memory address or nil pointer dereference [recovered]
> 	panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x10f5ac8]
> 
> goroutine 132 [running]:
> testing.tRunner.func1.2({0x126caa0, 0x2327760})
> 	/usr/local/go/src/testing/testing.go:1545 +0x1c4
> testing.tRunner.func1()
> 	/usr/local/go/src/testing/testing.go:1548 +0x360
> panic({0x126caa0?, 0x2327760?})
> 	/usr/local/go/src/runtime/panic.go:914 +0x218
> command-line-arguments_test.TestScaler(0x4000475860)
> 	/__w/keda/keda/tests/scalers/aws/aws_managed_prometheus/aws_managed_prometheus_test.go:143 +0x178
> testing.tRunner(0x4000475860, 0x15808b0)
> 	/usr/local/go/src/testing/testing.go:1595 +0xe8
> created by testing.(*T).Run in goroutine 1
> 	/usr/local/go/src/testing/testing.go:1648 +0x33c
> FAIL	command-line-arguments	0.491s
> FAIL
> ```
> 
> It looks that the test can't start for any reason, but the issue looks in the test script side, not in KEDA. Could you take a look?

Am unable to reproduce this. Test goes fine in my local when I try `GOOS="darwin" go test -v -tags e2e -timeout  2m scalers/aws/aws_managed_prometheus/aws_managed_prometheus_test.go` and I could be able to see the AMP workspace created and removed after the test is done. I set the `TF_AWS_ACCESS_KEY`, `TF_AWS_SECRET_KEY` and `TF_AWS_REGION`.. Is there a good way to replicate this?

File: pkg/scalers/aws/aws_sigv4.go
Comment (zroubalik) [inline pkg/scalers/aws/aws_sigv4.go:34]: I am sorry, I don't have enough time to dive into, but without knowing all the details, I am concerned about this part.

We should add an explanation why is this specific stuff neede, because it is not so obvious from the code. We shouldn't just add stuff to please semgrep imho.

File: pkg/scalers/aws/aws_sigv4.go
Comment (zroubalik) [inline pkg/scalers/aws/aws_sigv4.go:1]: add license header

File: pkg/scalers/aws/aws_sigv4.go
Comment (zroubalik) [inline pkg/scalers/aws/aws_sigv4.go:2]: Please add here a comment with a few sentences that describe what is this file about

File: pkg/scalers/aws/aws_sigv4.go
Comment (zroubalik) [inline pkg/scalers/aws/aws_sigv4.go:26]: add doc please

File: pkg/scalers/aws/aws_sigv4.go
Comment (zroubalik) [inline pkg/scalers/aws/aws_sigv4.go:47]: add doc

File: pkg/scalers/aws/aws_sigv4_test.go
Comment (zroubalik) [inline pkg/scalers/aws/aws_sigv4_test.go:1]: add license header please

**Comment by zroubalik** [comment]
> > Error logs from e2e test are really weird:
> > ```shell
> > === RUN   TestScaler
> >     aws_managed_prometheus_test.go:138: --- setting up ---
> > --- FAIL: TestScaler (0.42s)
> > panic: runtime error: invalid memory address or nil pointer dereference [recovered]
> > 	panic: runtime error: invalid memory address or nil pointer dereference
> > [signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x10f5ac8]
> > 
> > goroutine 132 [running]:
> > testing.tRunner.func1.2({0x126caa0, 0x2327760})
> > 	/usr/local/go/src/testing/testing.go:1545 +0x1c4
> > testing.tRunner.func1()
> > 	/usr/local/go/src/testing/testing.go:1548 +0x360
> > panic({0x126caa0?, 0x2327760?})
> > 	/usr/local/go/src/runtime/panic.go:914 +0x218
> > command-line-arguments_test.TestScaler(0x4000475860)
> > 	/__w/keda/keda/tests/scalers/aws/aws_managed_prometheus/aws_managed_prometheus_test.go:143 +0x178
> > testing.tRunner(0x4000475860, 0x15808b0)
> > 	/usr/local/go/src/testing/testing.go:1595 +0xe8
> > created by testing.(*T).Run in goroutine 1
> > 	/usr/local/go/src/testing/testing.go:1648 +0x33c
> > FAIL	command-line-arguments	0.491s
> > FAIL
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It looks that the test can't start for any reason, but the issue looks in the test script side, not in KEDA. Could you take a look?
> 
> Am unable to reproduce this. Test goes fine in my local when I try `GOOS="darwin" go test -v -tags e2e -timeout 2m scalers/aws/aws_managed_prometheus/aws_managed_prometheus_test.go` and I could be able to see the AMP workspace created and removed after the test is done. I set the `TF_AWS_ACCESS_KEY`, `TF_AWS_SECRET_KEY` and `TF_AWS_REGION`.. Is there a good way to replicate this?

even if you are not able to reproduce, the test shouldn't end with panic, we should add a check there and fire proper error anyway.

File: tests/scalers/aws/aws_managed_prometheus/aws_managed_prometheus_test.go
Comment (JorTurFer) [inline tests/scalers/aws/aws_managed_prometheus/aws_managed_prometheus_test.go:142]:
```suggestion
	assert.NoError(t, err, "aws prometheus workspace creation has failed")
```

File: tests/scalers/aws/aws_managed_prometheus/aws_managed_prometheus_test.go
Comment (JorTurFer) [inline tests/scalers/aws/aws_managed_prometheus/aws_managed_prometheus_test.go:163]:
```suggestion
	assert.NoError(t, err, "aws prometheus workspace deletion has failed")
```

File: tests/scalers/aws/aws_managed_prometheus_pod_identity/aws_managed_prometheus_pod_identity_test.go
Comment (JorTurFer) [inline tests/scalers/aws/aws_managed_prometheus_pod_identity/aws_managed_prometheus_pod_identity_test.go:148]:
```suggestion
	assert.NoError(t, err, "aws prometheus workspace deletion has failed")
```

File: tests/scalers/aws/aws_managed_prometheus_pod_identity/aws_managed_prometheus_pod_identity_test.go
Comment (JorTurFer) [inline tests/scalers/aws/aws_managed_prometheus_pod_identity/aws_managed_prometheus_pod_identity_test.go:127]:
```suggestion
	assert.NoError(t, err, "aws prometheus workspace creation has failed")
```

File: pkg/scalers/aws/aws_sigv4.go
Comment (JorTurFer) [reply pkg/scalers/aws/aws_sigv4.go:34]: > Do we really need to add some constant or do hash on empty object?

We need to sign the request because giving an empty string (not a hashed empty string) fails in the backend as not signed request, so the answer is yes, we need to provide the value (I tested it). 
The question is: Should we calculate the hash from empty or adding a constant with the value? I think that we can place it as constant if we document it well. I mean, adding the constant as magic string without any other info is a problem, but I guess that if we document it there isn't any problem with it. We are already using constants with magic string for other things [like scopes](https://github.com/kedacore/keda/blob/main/pkg/scalers/azure_pipelines_scaler.go#L37)

**Comment by JorTurFer** [comment]
> even if you are not able to reproduce, the test shouldn't end with panic, we should add a check there and fire proper error anyway.

I think that my previous code suggestion was wrong and that's why the test panic before any assertion error xD

**Comment by sguruvar** [comment]
> > even if you are not able to reproduce, the test shouldn't end with panic, we should add a check there and fire proper error anyway.
> 
> I think that my previous code suggestion was wrong and that's why the test panic before any assertion error xD

Looks like permission is missing to create the AMP workspace: `operation error amp: CreateWorkspace, https response error StatusCode: 40***, RequestID: defa5098-e960-4***4-b684-a0c46dfdf6***d, AccessDeniedException: User: arn:aws:iam::***:user/e***e-test-user is not authorized to perform: aps:CreateWorkspace on resource: arn:aws:aps:***:***:/workspaces`

**Comment by JorTurFer** [comment]
It seems that it works, so once all the commits are addressed, we can merge this

File: pkg/scalers/aws/aws_sigv4.go
Comment (sguruvar) [reply pkg/scalers/aws/aws_sigv4.go:60]: I added as a const magic string with doc of why we're doing.

**Comment by sguruvar** [comment]
> It seems that it works, so once all the commits are addressed, we can merge this

Thanks.. I believe all teh comments are addressed. Can you help merge

File: pkg/scalers/aws/aws_sigv4.go
Comment (zroubalik) [inline pkg/scalers/aws/aws_sigv4.go:73]:
```suggestion
// parseAwsAMPMetadata parses the data to get the AWS sepcific auth info and metadata
```

File: pkg/scalers/aws/aws_sigv4.go
Comment (zroubalik) [reply pkg/scalers/aws/aws_sigv4.go:60]: @sguruvar could you please update the comment? we don't need to mention semgrep, but we should add a comment why we need use that hash value etc. Thanks!

Comment (sguruvar) [reply pkg/scalers/aws/aws_sigv4.go]: > @sguruvar could you please update the comment? we don't need to mention semgrep, but we should add a comment why we need use that hash value etc. Thanks!

done

**Review by zroubalik — APPROVED** [review]

**Pull request merged**
